test/helpers.lua

```diff
@@ -106,20 +106,32 @@ local uname = (function()
   end)
 end)()
 
-local function tmpname()
-  local fname = os.tmpname()
-  if uname() == 'Windows' and fname:sub(1, 2) == '\\s' then
-    -- In Windows tmpname() returns a filename starting with
-    -- special sequence \s, prepend $TEMP path
-    local tmpdir = os.getenv('TEMP')
-    return tmpdir..fname
-  elseif fname:match('^/tmp') and uname() == 'Darwin' then
-    -- In OS X /tmp links to /private/tmp
-    return '/private'..fname
-  else
-    return fname
-  end
-end
+local tmpname = (function()
+  local seq = 0
+  local tmpdir = os.getenv('TMPDIR') and os.getenv('TMPDIR') or os.getenv('TEMP')
+  -- Is $TMPDIR defined local to the project workspace?
+  local in_workspace = tmpdir and string.find(tmpdir, 'Xtest')
+  return (function()
+    if in_workspace then
+      seq = seq + 1
+      local fname = tmpdir..'/nvim-test-lua-'..seq
+      io.open(fname, 'w'):close()
+      return fname
```

Comment:
Lua's `os.tmpname` has numerous problems, in particular it doesn't respect $TMPDIR nor any other env var. The `seq = seq + 1` hack works fine since we know the temp dir is the workspace-local directory set by RunTests.cmake.